### ISSUE TITLE
need design for actor-related objects to cache in the session

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#598]
**Reported by @cristianbogdan on 20 Jul 2008 09:38 UTC**
from a chat with Rudi: there seems to be a need to not just keep in 
session the fields of an actor but also the fields of certain related 
objects. like best.Student needs its person of type general.Person... 
some possibilities:
1) indicate explicitly the pointers of objects that you want cached in 
the session
like person= session ptr general.Person
2) lazily cascade by default on all pointers
so if you see
actor(best.Student).ptr1.ptr2.ptr3, cache all members of 
actor(best.Student).ptr1, actor(best.Student).ptr1.ptr2, 
actor(best.Student).ptr1.ptr2.ptr3

but before actor(best.Student).ptr1 is mentioned in a query (or so), 
nothing from the ptr1 (except the primary key) is cached.

in this case if <mak:logout actor="best.Student"/> sees 
actor_best_Student_ptr1 it would remove actor_best_Student_ptr1*  also 
the special session variables created by the MDD of ptr1. same for 
ptr1.ptr2 and so on.

there would be huge difficulties of keeping all these up to date and i 
guess you already have these difficulties in karamba. because if i 
change ptr1.field then i have to check if ptr1 is not in some session, 
then update the session as well. if i change an actor field, it is 
somewhat easier.

Migrated-From: http://trac.makumba.org/ticket/1052

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 20 Jul 2008 10:13 UTC**
(In reply to comment #0)
> 1) indicate explicitly the pointers of objects that you want cached in 
> the session
> like person= session ptr general.Person
> 2) lazily cascade by default on all pointers
> so if you see
> actor(best.Student).ptr1.ptr2.ptr3, cache all members of 
> actor(best.Student).ptr1, actor(best.Student).ptr1.ptr2, 
> actor(best.Student).ptr1.ptr2.ptr3
> 
> but before actor(best.Student).ptr1 is mentioned in a query (or so), 
> nothing from the ptr1 (except the primary key) is cached.

one interesting version of this (let's call it 2.1) is to inline only over fixed pointers. then we know that at least these will never change, making the session easier to maintain. however, the fields of the objects identified by these pointers might change...
 
> there would be huge difficulties of keeping all these up to date and i 
> guess you already have these difficulties in karamba. because if i 
> change ptr1.field then i have to check if ptr1 is not in some session, 
> then update the session as well. if i change an actor field, it is 
> somewhat easier.

this actually brings me to

3) do not cache any members or functions...

it seems to me that the session variable euphoria that currently exists in Karamba can be addressed easily with actor(type).ptr.ptr.member or actor(type).ptr.function(), besides there is the possibility of passing parameters to such functions, etc.

in other words i think that the 'larger' sessions computed in karamba are an effect of not being able to standardize code, which currently exists with actors and functions.

after all, an actor(type).ptr.ptr.member only introduces one projection, one FROM and one WHERE (see bug 1054), and if the same actor(type) is used in the same query, the FROM and the WHERE will be shared (see bug 1053), only one more projection will be added. 

4) is a relaxed version of (3) in regard to complex functions, that might need subqueries, which are not available in OQL. those would need to be done separately, and for that (3) be relaxed...

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 24 Jul 2008 09:56 UTC**
there is also a need to access the session data from JSP EL ${} expressions. actually besides BL (which will be simplified by functions anyway) this is a very important use of the session data.

my current proposal for this, which works with proposal (3) but also with (1) and (2) is to define via the makumba controller a clever map called e.g. actor. then:

${actor.general_Person_name} will lead to a get("general_Person_name") in the clever Map, which would then access session data and/or do queries to find the result. 

several ${actor_Type_*} from a page, if (like in approach (3)) lead to a query can be grouped in the same query so there would be only one query per actor per page (maybe actor object's functions would be computed in separate queries though, to avoid need for subqueries)

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 24 Jul 2008 13:44 UTC**
one more idea here, how about caching values between explicit and implicit authorization?

let's say in MC.properties there is
authorize%/path={actor(SomeType).someComplicatedFunction() }

and then in the canEdit of a mak:editForm under /path the same function is invoked again. this is not unlikely because in MC.properties you want to protect a whole part of the site, including this editing page. of course you could skip the canEdit() but i think it's really good to have it there for the MDD to be self-contained.

in that case it looks like caching the function result just during this process is a good thing because it's unlikely that it changes its value during authorization, before any BL actually starts.

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 11 Aug 2009 11:27 UTC**
I think one good page to look at is:

http://private.best.eu.org/itc/sessionAttributes.jsp

what would be even better is to have some statistics regarding the number of pages in which each of those session attributes is needed.

right now, my thinking goes like this:

- if the page is running a query anyway, then adding a projection to the specific field of a type that can be retrieved via the actor shouldn't have a too big impact on performance. it might have if there's a need to do a lot of cascading joins (like best.internal.Student bis join best.Student bs join bs.person p where bis.student = bs and bs.person = p), but I'm not even sure about that

- in any case however, this additional projection and FROM should be added to the query of the page, and not generate another query

- there's a need for supporting hashmap/vector session attributes, or perhaps rather to be able to mimic the current hashmaps of the kind _sessVec_MemberOf_GeneralTeam which allow to do EL like _sessVec_MemberOf_GeneralTeam['teamPtrExternalForm'] if I'm not mistaken. I don't know to what extent the clever map could be used for that purpose? anyway it sounds more like a function to me, that could then be used like actor(Person).isMemberOfGeneralTeam(teamPtr) - yet that would require a lot of refactoring of K! (but I think this should not be a criteria to adopt good solutions.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 27 Apr 2010 20:18 UTC**
the thick session is commented out and if we decide to drop it we should also remove the references to this bug from the documentation

http://www.makumba.org/page/QueryFragments#section-QueryFragments-ActorFunctions

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**
